services/management/ia/rules_service.go

```diff
@@ -255,11 +256,88 @@ func (s *RulesService) ListAlertRules(ctx context.Context, req *iav1beta1.ListAl
 		return nil, status.Errorf(codes.FailedPrecondition, "%v.", services.ErrAlertingDisabled)
 	}
 
-	res, err := s.getAlertRules()
+	if req.Page == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Page should be set")
+	}
+	if req.Page.Size <= 0 || req.Page.Index < 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "Page size (%d) should be positive number and "+
+			"page index (%d) should be non-negative number", req.Page.Size, req.Page.Index)
+	}
+
+	res, pageTotals, err := s.getAlertRulesPage(req.Page)
 	if err != nil {
 		return nil, err
 	}
-	return &iav1beta1.ListAlertRulesResponse{Rules: res}, nil
+	return &iav1beta1.ListAlertRulesResponse{Rules: res, Totals: pageTotals}, nil
+}
+
+func (s *RulesService) convertAlertRules(rules []*models.Rule, channels []*models.Channel) ([]*iav1beta1.Rule, error) {
+	templates := s.templates.getTemplates()
+
+	res := make([]*iav1beta1.Rule, 0, len(rules))
+	for _, rule := range rules {
+		template, ok := templates[rule.TemplateName]
+		if !ok {
+			s.l.Warnf("template %s used by rule %s doesn't exist, skipping that rule", template.Name, rule.ID)
+			continue
+		}
+
+		r, err := convertRule(s.l, rule, template, channels)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		res = append(res, r)
+	}
+
+	return res, nil
+}
+
+// getAlertRulesPage returns a page with list of available alert rules.
+func (s *RulesService) getAlertRulesPage(page *iav1beta1.Page) ([]*iav1beta1.Rule, *iav1beta1.PageTotals, error) {
+	var rules []*models.Rule
+	var channels []*models.Channel
+	var totalItems int
+	errTx := s.db.InTransaction(func(tx *reform.TX) error {
+		var err error
+		rules, err = models.FindRulesOnPage(tx.Querier, int(page.Index), int(page.Size))
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		totalItems, err = models.CountRules(tx.Querier)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		channelsIDs := make(map[string]struct{}, len(rules))
+		for _, rule := range rules {
+			for _, id := range rule.ChannelIDs {
+				channelsIDs[id] = struct{}{}
+			}
+		}
+
+		channels, err = models.FindChannelsByIDs(tx.Querier, utils.SetToSlice(channelsIDs))
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+	if errTx != nil {
+		return nil, nil, errors.WithStack(errTx)
+	}
+
+	res, err := s.convertAlertRules(rules, channels)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	totals := &iav1beta1.PageTotals{
+		TotalItems: int32(totalItems),
+		TotalPages: int32(totalItems) / page.Size,
```

Comment:
Truncation could happen. You need to round up.
101(items)/20(page size) => should be 6, not 5

Reply:
done, thanks.